awsutil/awsutil.go

```diff
@@ -0,0 +1,124 @@
+package awsutil
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/mitchellh/mapstructure"
+)
+
+// Some AWS destinations are using SecretAccessKey instead of accessKey
+type SessionConfig struct {
+	Region           string         `mapstructure:"region"`
+	AccessKeyID      string         `mapstructure:"accessKeyID"`
+	AccessKey        string         `mapstructure:"accessKey"`
+	SecretAccessKey  string         `mapstructure:"secretAccessKey"`
+	RoleBasedAuth    bool           `mapstructure:"roleBasedAuth"`
+	IAMRoleARN       string         `mapstructure:"iamRoleARN"`
+	ExternalID       string         `mapstructure:"externalID"`
+	WorkspaceID      string         `mapstructure:"workspaceID"`
+	Endpoint         *string        `mapstructure:"endpoint"`
+	S3ForcePathStyle *bool          `mapstructure:"s3ForcePathStyle"`
+	DisableSSL       *bool          `mapstructure:"disableSSL"`
+	Service          string         `mapstructure:"service"`
+	Timeout          *time.Duration `mapstructure:"timeout"`
+}
+
+// CreateSession creates a new AWS session using the provided config
+func CreateSession(config *SessionConfig) (*session.Session, error) {
+	var (
+		awsCredentials *credentials.Credentials
+		err            error
+	)
+	if config.RoleBasedAuth {
+		awsCredentials, err = createCredentialsForRole(config)
+	} else if config.AccessKey != "" && config.AccessKeyID != "" {
+		awsCredentials, err = credentials.NewStaticCredentials(config.AccessKeyID, config.AccessKey, ""), nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	return session.NewSession(&aws.Config{
+		HTTPClient:                    getHttpClient(config),
+		Region:                        aws.String(config.Region),
+		CredentialsChainVerboseErrors: aws.Bool(true),
+		Credentials:                   awsCredentials,
+		Endpoint:                      config.Endpoint,
+		S3ForcePathStyle:              config.S3ForcePathStyle,
+		DisableSSL:                    config.DisableSSL,
+	})
+}
+
+// NewSimpleSessionConfig creates a new session config using the provided config map
+func NewSimpleSessionConfig(config map[string]interface{}, serviceName string) (*SessionConfig, error) {
+	if config == nil {
+		return nil, errors.New("config should not be nil")
+	}
+	sessionConfig := SessionConfig{}
+	if err := mapstructure.Decode(config, &sessionConfig); err != nil {
+		return nil, fmt.Errorf("unable to populate session config using destinationConfig: %w", err)
+	}
+
+	if !isRoleBasedAuthFieldExist(config) {
+		sessionConfig.RoleBasedAuth = sessionConfig.IAMRoleARN != ""
+	}
+
+	if sessionConfig.IAMRoleARN == "" {
+		sessionConfig.RoleBasedAuth = false
+	}
```

Comment:
This set of conditions seem a bit confusing to me. Perhaps a test to verify the expectations here might be useful.